src/libsystemd-network/dhcp6-option.c

```diff
@@ -410,14 +404,64 @@ int dhcp6_option_parse(
         return 0;
 }
 
-int dhcp6_option_parse_status(DHCP6Option *option, size_t len) {
-        DHCP6StatusOption *statusopt = (DHCP6StatusOption *)option;
+int dhcp6_option_parse_status(const uint8_t *data, size_t data_len, char **ret_status_message) {
+        assert(data);
 
-        if (len < sizeof(DHCP6StatusOption) ||
-            be16toh(option->len) + offsetof(DHCP6Option, data) < sizeof(DHCP6StatusOption))
-                return -ENOBUFS;
+        if (data_len < sizeof(uint16_t))
+                return -EBADMSG;
+
+        if (ret_status_message) {
+                char *msg;
 
-        return be16toh(statusopt->status);
+                /* The status message MUST NOT be null-terminated. See section 21.13 of RFC8415. */
+                msg = memdup_suffix0(data + sizeof(uint16_t), data_len - sizeof(uint16_t));
+                if (!msg)
+                        return -ENOMEM;
```

Comment:
shouldn't this check for embedded NUL bytes, and do something about it? i.e. either fail, or replace by space, or at least log?

Reply:
Well, I think it is not necessary, this is just a status message and it is used in log message.

Reply:
i think even for log messages we should probably do some basic sanitation...

Actually, given we dump this message potentially to the tty we robably should filter control characters and invalid utf8?

Maybe treat all the same here, and instead of using memdup_suffix(), just use cescape_length()? If the string is nicely put together this should not transform anything. But if it contains NUL bytes, control characters or bad utf8 it will all catch that and escape it, so that we never output something unreadable and dangerious...